src/java/org/apache/cassandra/index/sai/plan/QueryController.java

```diff
@@ -133,22 +133,16 @@ public List<DataRange> dataRanges()
         return ranges;
     }
 
-    /**
-     * @return indexed {@code IndexContext} if index is found; otherwise return non-indexed {@code IndexContext}.
-     */
-    public IndexContext getContext(RowFilter.Expression expression)
+    public StorageAttachedIndex indexFor(RowFilter.Expression expression)
     {
         Set<StorageAttachedIndex> indexes = cfs.indexManager.getBestIndexFor(expression, StorageAttachedIndex.class);
+        return indexes.isEmpty() ? null : indexes.iterator().next();
+    }
 
-        return indexes.isEmpty() ? new IndexContext(cfs.getKeyspaceName(),
-                                                    cfs.getTableName(),
-                                                    cfs.metadata().partitionKeyType,
-                                                    cfs.getPartitioner(),
-                                                    cfs.getComparator(),
-                                                    expression.column(),
-                                                    IndexTarget.Type.VALUES,
-                                                    null)
-                                 : indexes.iterator().next().getIndexContext();
+    public AbstractAnalyzer analyzerFor(RowFilter.Expression expression)
+    {
+        StorageAttachedIndex index = indexFor(expression);
+        return index == null ? new NoOpAnalyzer() : index.analyzer();
```

Comment:
I know the `NoOpAnalyzer` makes some other parts of the code simpler perhaps, but the object creations are still kind of bothersome when we don't transform the value.

Reply:
I agree. I've been debating this because I don't like the extra code for the noop analyzer. As it stands all string handling goes through the analyzer path, where a good number probably don't use an analyzer even on indexed strings.